### PR TITLE
feat: add Distill pipline 

### PR DESCRIPTION
**Pull request by yisunlp**
## Add distillation training support for full-precision and quantized students

This PR adds a first-class `Distill` compressor to AngelSlim. It supports teacher-student distillation for both full-precision students and quantized QAT-style students, and keeps the existing PTQ/QAT paths unchanged.

The feature introduces an independent teacher model, a distillation trainer, stable CausalLM-aligned KD loss computation, SFT-style prompt masking for chat data, and example configs for two common workflows:

- Distilling a smaller full-precision student from a larger full-precision teacher.
- Distilling a quantized student from a full-precision teacher.

### New Distill compressor

`angelslim/compressor/distill/distill.py` and `angelslim/compressor/distill/trainer.py`

- Register `Distill` in `CompressorFactory`.
- Load a separate full-precision teacher from `compression.Distill.teacher_model_path`.
- Support `student_type: fp` and `student_type: quantized`.
- Support `trainable_parameters: all` and `trainable_parameters: quant`.
- Use HuggingFace `Seq2SeqTrainer` arguments from `compression.Distill.hf_args`.
- Save full-precision students with HuggingFace `save_pretrained`.
- Reuse QAT conversion and save paths for quantized students.

### Distillation loss

`angelslim/compressor/qat/plugins/distill_loss.py`

- Align KD tokens with HuggingFace CausalLM next-token shifting.
- Compute KD only on valid labels after prompt/padding masks.
- Support `kl`, `rkl`, `kd`, `cakld`, `mse`, `kl_top`, and `rkl_top` variants.

### Dataset handling

`angelslim/data/text_dataset.py` and `angelslim/data/qat_dataset.py`

- Add chat-style SFT masking for JSONL data.
- When `is_sft_data: true`, prompt tokens are masked with `-100`, and only the final assistant response contributes to LM and KD losses.
- Preserve `input_ids`, `attention_mask`, and `labels` in the QAT dataset wrapper.
- Skip samples with no valid supervised tokens to avoid NaN loss.

### Config parsing and engine integration

`angelslim/utils/config_parser.py`, `angelslim/engine.py`, and `tools/run.py`

- Add `DistillTrainingConfig`.
- Parse `compression.name: Distill` and `compression.Distill` fields.
- Require quantization config only when `student_type: quantized`.
- Normalize `device_map: None` / `distributed` for DeepSpeed-managed model placement.
- Pre-warm HuggingFace DeepSpeed config before model loading when needed.

### 1. Full-precision student distillation

Config:

```text
configs/qwen3/distill/fp/qwen3-1_7b_fp_distill_from_qwen3-4b_zero2.yaml
```

This example distills a Qwen3-1.7B full-precision student from a Qwen3-4B full-precision teacher.

```bash
torchrun --nproc_per_node=2 \
  tools/run.py \
  -c configs/qwen3/distill/fp/qwen3-1_7b_fp_distill_from_qwen3-4b_zero2.yaml
```

### 2. Full-precision teacher to quantized student

Config:

```text
configs/qwen3/distill/w4a8_fp8/qwen3-4b_w4a8_fp8_distill_zero2.yaml
```

This example distills a W4A8-FP8 Qwen3-4B student from a full-precision Qwen3-4B teacher. The quantized student reuses QAT learnable-scale plugins and trains quantization parameters.

```bash
torchrun --nproc_per_node=2 \
  tools/run.py \
  -c configs/qwen3/distill/w4a8_fp8/qwen3-4b_w4a8_fp8_distill_zero2.yaml
```

## Rollout Data Processing

The PR includes a reproducible data-construction guide and script under:

```text
dataset/qwen3_4b_rollout_10k/
```

Only the construction script and README are included. Generated JSONL files are ignored by `.gitignore` and are not committed.

Workflow:

1. Sample prompts from public instruction datasets.
2. Apply prompt deduplication and length filtering.
3. Filter prompts that explicitly mention common benchmark names such as MMLU, GSM8K, ARC, HellaSwag, and Winogrande.
4. Generate teacher responses with vLLM.
5. Remove empty, too-short, too-long, or refusal-template responses.
6. Save chat-style JSONL rows compatible with `TextDataset(is_sft_data=True)`.

The generated format is:

```json
{
  "id": "reasoning_math_000001",
  "source": "TIGER-Lab/MathInstruct",
  "category": "reasoning_math",
  "rollout_model": "Qwen/Qwen3-4B",
  "thinking": true,
  "prompt_tokens": 128,
  "response_tokens": 512,
  "total_tokens": 640,
  "messages": [
    {"role": "user", "content": "..."},
    {"role": "assistant", "content": "..."}
  ]
}
```

## Benchmark Results

The following benchmark compares the Qwen3-1.7B base model with the Qwen3-1.7B model distilled from a Qwen3-4B teacher. PPL is not included in the comparison below.

| Group | Task | Base | Distilled | Delta | Samples |
|---|---:|---:|---:|---:|---:|
| General | PIQA | 0.6638 | 0.7383 | +0.0745 | 1838 |
| General | ARC Easy | 0.8930 | 0.8912 | -0.0018 | 570 |
| General | ARC Challenge | 0.7258 | 0.7224 | -0.0034 | 299 |
| General | HellaSwag | 0.5908 | 0.6257 | +0.0349 | 10042 |
| General | Winogrande | 0.5446 | 0.5304 | -0.0142 | 1267 |
| General | MMLU | 0.5291 | 0.5096 | -0.0195 | 14042 |
| Reasoning | GSM8K | 0.7991 | 0.7612 | -0.0379 | 1319 |
| Reasoning | MATH subset | 0.6081 | 0.6040 | -0.0041 | 500 |
| Reasoning | BBH subset | 0.7000 | 0.8000 | +0.1000 | 250 |

## Configs / Scripts / Docs

- `configs/qwen3/distill/fp/qwen3-1_7b_fp_distill_from_qwen3-4b_zero2.yaml`
- `configs/qwen3/distill/w4a8_fp8/qwen3-4b_w4a8_fp8_distill_zero2.yaml`
- `configs/qwen3/distill/w4a8_fp8/ds_config_zero2.json`
- `scripts/distill/run_distill_for_qwen_4b_zero2.sh`
- `docs/source/features/distill/index.md`
- `dataset/qwen3_4b_rollout_10k/README.md`
- `dataset/qwen3_4b_rollout_10k/prepare_qwen3_rollout_data.py`

### CODE REVIEW

File: configs/qwen3/distill/fp/qwen3-1_7b_fp_distill_cakld_from_qwen3-4b_zero2.yaml [hunk 1]
```diff
@@ -0,0 +1,55 @@
+global:
```
Comment (yghstill) [inline configs/qwen3/distill/fp/qwen3-1_7b_fp_distill_cakld_from_qwen3-4b_zero2.yaml:1]: 文件命名中最好包含蒸馏策略，方便后续扩展不同的蒸馏算法

**Review by yghstill — APPROVED** [review]
